docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,389 @@
+
+.. code-block:: ql
+
+     TaintTracking::localTaint(source, sink)
+
+
+Using local sources
+~~~~~~~~~~~~~~~~~~~
+
+When asking for local data flow or taint propagation between two expressions as above, you would normally constrain the expressions to be relevant to a certain investigation.
+The next section will give some concrete examples, but there is a more abstract concept that we should call out explicitly, namely that of a local source.
+
+A local source is a data-flow node with no local data flow into it.
+As such, it is a local origin of data flow, a place where a new value is created.
+This includes parameters (which only receive global data flow) and most expressions (because they are not value-preserving).
+The class ``LocalSourceNode`` represents data-flow nodes that are also local sources.
+It comes with a useful member predicate ``flowsTo(DataFlow::Node node)``, which holds if there is local data flow from the local source to ``node``.
+
+Examples
+~~~~~~~~
+
+This query finds the filename argument passed in each call to ``File.open``:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.ApiGraphs
+    
+    from DataFlow::CallNode call
+    where call = API::getTopLevelMember("File").getAMethodCall("open")
+    select call.getArgument(0)
+
+Notice the use of the ``API`` module for referring to library methods.
+For more information, see ":doc:`Using API graphs in Ruby <using-api-graphs-in-ruby>`."
+
+Unfortunately this will only give the expression in the argument, not the values which could be passed to it.
+So we use local data flow to find all expressions that flow into the argument:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.ApiGraphs
+    
+    from DataFlow::CallNode call, DataFlow::ExprNode expr
+    where
+      call = API::getTopLevelMember("File").getAMethodCall("open") and
+      DataFlow::localFlow(expr, call.getArgument(0))
+    select call, expr
+
+Many expressions flow to the same call.
+If you run this query, you may notice that you get several data-flow nodes for an expression as it flows towards a call (notice repeated locations in the ``call`` column).
+We are mostly interested in the "first" of these, what might be called the local source for the file name.
+To restrict attention to such local sources, and to simultaneously make the analysis more performant, we have the QL class ``LocalSourceNode``.
+We could demand that ``expr`` is such a node:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.ApiGraphs
+    
+    from DataFlow::CallNode call, DataFlow::ExprNode expr
+    where
+      call = API::getTopLevelMember("File").getAMethodCall("open") and
+      DataFlow::localFlow(expr, call.getArgument(0)) and
+      expr instanceof DataFlow::LocalSourceNode
+    select call, expr
+
+However, we could also enforce this by casting.
```

Comment:
Personally I'd just change the type of `expr` to `DataFlow::LocalSourceNode`. Would that be any clearer?

Reply:
I think it would be, but if we're going to do that, we should also do it in the Python docs.